### ISSUE TITLE
[Research Story]: Audit Build Settings for Performance Improvements

### ISSUE THREAD

**Issue by cherdman** [magento/pwa-studio#1623]
**What?**
Audit the Webpack build configuration to ensure that it builds the smallest and fastest possible app code.

**Why?**
We need to understand performance issues in relation to our build process. Webpack is very configurable and it's possible we could achieve great performance gains here.

**How?**
We will play with build settings, rebuild, measure, experiment, and repeat, logging our results along the way.

We'll determine the best parameters for:
  - code splitting
  - optimization and minification settings
  - deduplication plugins
  - browser and device support without unnecessary polyfills
  - reducing built times for development and CI
  - tree shaking and dead code elimination

**Acceptance Criteria**
- Write-up of audit results and proposed changes
- Actions and next steps created in the form of GitHub issues

**Comment by revanth0212** [comment]
1. Separating out runtime code from actual code into a separate chunk will only load the actual code when a new build is generated. This step will create 3 chunk as opposed to 2 as of today, namely, client (which contains business logic), runtime (which contains webpack related logic) and vendors (which contains all the node_modules logic needed by client).
2. Adding hash to chunk file names, will cache the files on the client end and only request if the index.html file references a new filename that is not cached in the past.

**Comment by zetlen** [comment]
1. **Runtime chunk** -- I've experimented with this before and not found a great improvement, but there are a lot of settings to tweak and I may have gotten it wrong. I'd love to see a demo that improves performance.
2. **Minification** -- I'm very interested in the changes we might see from `babel-minify`, but I'm concerned that there would be a regression in our tree-shaking. Right now, the TerserWebpackPlugin does dead code elimination based on logic it determines inside each file. I'm not sure that global DCE would work the same way. It sounds like we'd have an overall reduction in bundle size, which is great, but I want to understand how it would behave in more detail.
3. **CSS Extraction** -- Since many of our modules load asynchronously, I would want to be sure that the CSS for them loads asynchronously as well. Right now, leaving the CSS as part of the bundle is the easiest way of ensuring that. I would approve of extracted CSS, but only if a) the CSS is only for the "critical path", components that will show up in a VERY early render, and b) the CSS is inlined into the application shell itself. I know this makes it harder to do cache headers, but in a mature production application the same cache headers would apply to the bundles-including-stylesheets anyway.
4. **Multiple Entry Chunks** -- I'd just need to see how this works with zero SSR and dynamic routes for products and categories. I'm not sure how it would work, but I'm totally willing to see it. The one thing I'd be cautious of is duplicating vendor code between chunks--it seems necessary to have at least one common vendor bundle to avoid this.

Thanks, @revanth0212 ! I know this is a lot of followup, but mainly I want to say that I really appreciate the depth of research.

**Comment by revanth0212** [comment]
Thanks for that @zetlen. I'll work on the numbers to get more insights into the suggestions I made earlier.

> 1. **Runtime chunk** -- I've experimented with this before and not found a great improvement, but there are a lot of settings to tweak and I may have gotten it wrong. I'd love to see a demo that improves performance.

The runtime chunk will change with every run of webpack even if the app code has not changed. Separating it out will not make the UI fetch client and vendors code because they haven't changed making the refresh look faster.

> 2\. **Minification** -- I'm very interested in the changes we might see from `babel-minify`, but I'm concerned that there would be a regression in our tree-shaking. Right now, the TerserWebpackPlugin does dead code elimination based on logic it determines inside each file. I'm not sure that global DCE would work the same way. It sounds like we'd have an overall reduction in bundle size, which is great, but I want to understand how it would behave in more detail.

Honestly, I am not sure how did that work but as most of the discoveries, it was by accident. I was toying with a bunch of minification libraries and ended up with this guy. When I read about it, it sounded just like Terser. According to a bunch of blog posts and benchmarks, Terser does the best job of Minification and Uglification. Regarding Dead Code Elimination, it is built into webpack and according to their docs, it does Dead Code analysis at the global level, not just file level. There is more we can do with DCE with webpack but that includes a lof of manul work where we have to tag each line of code with `/*#__PURE__*/` letting webpack know that it can analyze the code without side effects checking.

> 3\. **CSS Extraction** -- Since many of our modules-load asynchronously, I would want to be sure that the CSS for them loads asynchronously as well. Right now, leaving the CSS as part of the bundle is the easiest way of ensuring that. I would approve of extracted CSS, but only if a) the CSS is only for the "critical path", components that will show up in a VERY early render, and b) the CSS is inlined into the application shell itself. I know this makes it harder to do cache headers, but in a mature production application, the same cache headers would apply to the bundles-including-stylesheets anyway.

I may have lost you here for a bit but what I understand is, make sure there no race conditions. Also, make sure the main page that is being requested has its CSS available during first paint.

This might be hard but I am assuming possible because if we can have multiple chunks, we can have multiple CSS files for each for them. But again that's an overhead for us maintainers and webpack at the same time.

Let me know if I am missing something here.

> 4\. **Multiple Entry Chunks** -- I'd just need to see how this works with zero SSR and dynamic routes for products and categories. I'm not sure how it would work, but I'm totally willing to see it. The one thing I'd be cautious of is duplicating vendor code between chunks--it seems necessary to have at least one common vendor bundle to avoid this.

If I understand your point correctly, my solution is to have different HTML files for different routes serving different bundle and CSS files. This is a long shot but will increase network performance and memory footprint of the app.

Regarding vendor bundle, as of today, we can go with a single bundle that has all the vendors logic for the whole app and with client caching enabled, we don't need to download it on every refresh or every route change. Every route change will only download the route logic bundle and use the vendor and runtime bundle that it cached previously.

Basically, everything rotates around client-side caching.

**Comment by revanth0212** [comment]
With Client Side caching is enabled and runtime bundle is separated and App code has been modified, a new client build is generated. Meaning when user refreshes or does a page transition after new deployment, only 443KB (New client.js) of data is downloaded instead of 1.08MB of data (New client.js and old vendors.js) as of today.

![image](https://user-images.githubusercontent.com/35203638/64725796-f3180300-d49a-11e9-8a6a-a0fd94cc585f.png)

First Time (Client and Vendors both are downloaded)

![image](https://user-images.githubusercontent.com/35203638/64725478-48074980-d49a-11e9-9e0b-6367e1df4a19.png)

Second Time - No code change only refresh (No New Bundle is downloaded)

![image](https://user-images.githubusercontent.com/35203638/64725550-6e2ce980-d49a-11e9-84a1-02b147e28254.png)

Third Time: App code changed (Only the Client is downloaded)

![image](https://user-images.githubusercontent.com/35203638/64725656-a8968680-d49a-11e9-8410-828a0d1428e0.png)

**Comment by revanth0212** [comment]
Also, as of today the service worker is not minified and cached. Working on that along with moving the service worker register logic to a cached file of its own so it won't be downloaded more than once (mainly because the register never changes). It doesn't make sense to leave it in the bundle as well since the bundle can change over time but the register logic doesn't.

**Comment by supernova-at** [comment]
@revanth0212 can we close this research story?

**Comment by revanth0212** [comment]
> @revanth0212 can we close this research story?

Yes we can since all the findings are addressed in #1691
